emission/tests/modellingTests/TestBackwardsCompat.py

```diff
@@ -0,0 +1,207 @@
+        return user_input_pred_list, this_cluster_size
+
+    def testRandomTripsWithinTheSameThreshold(self):
+        label_data = {
+            "mode_confirm": ['walk', 'bike', 'transit'],
+            "purpose_confirm": ['work', 'home', 'school'],
+            "replaced_mode": ['drive']
+        }
+
+        n = 60
+        trips = etmm.generate_mock_trips(
+            user_id="joe", 
+            trips=n, 
+            origin=(0, 0), 
+            destination=(1, 1), 
+            label_data=label_data, 
+            threshold=0.001,  # ~ 111 meters in degrees WGS84
+        )
+
+        # These fields should ignored for the first round, but are extracted anyway
+        # So let's fill them in with dummy values
+        for t in trips:
+            t["data"]["distance"] = 1000
+            t["data"]["duration"] = 10
+
+        train = trips[0:50]
+        test = trips[50:60]
+
+        model_config = {
+            "metric": "od_similarity",
+            "similarity_threshold_meters": 500,      # meters,
+            "apply_cutoff": False,
+            "incremental_evaluation": False
+        }
+        new_model = eamtg.GreedySimilarityBinning(model_config)
+        new_model.fit(train)
+
+        old_builder = oursim.similarity(train, 500,
+            shouldFilter=False, cutoff=False)
+        old_builder.fit()
+
+        self.assertEqual(len(old_builder.bins), len(new_model.bins),
+            f"old bins = {old_builder.bins} but new_bins = {new_model.bins}")
+
+        self.assertEqual(len(old_builder.bins), 1,
+            f"all trips within threshold, so expected one bin, found {len(old_builder.bins)}")
+
+        old_user_inputs = eamtb.create_user_input_map(train, old_builder.bins)
+        old_location_map = eamtb.create_location_map(train, old_builder.bins)
+        old_cluster_sizes = {k: len(old_location_map[k]) for k in old_location_map}
+
+        for test_trip in test:
+            new_results, new_n = new_model.predict(test_trip)
+            old_results, old_n = TestBackwardsCompat.old_predict_with_n(test_trip,
+                old_location_map, old_user_inputs, old_cluster_sizes, 500)
+
+            self.assertEqual(old_n, new_n,
+                f"for test trip {test_trip} old n = {old_n} and new_n = {new_n}")
+
+            self.assertEqual(old_results, new_results,
+                f"for test trip {test_trip} old result = {old_results} and new result = {new_results}")
+
+    def testRandomTripsOutsideTheSameThreshold(self):
+        label_data = {
+            "mode_confirm": ['walk', 'bike', 'transit'],
+            "purpose_confirm": ['work', 'home', 'school'],
+            "replaced_mode": ['drive']
+        }
+
+        n = 60
+        trips = etmm.generate_mock_trips(
+            user_id="joe", 
+            trips=n, 
+            origin=(0, 0), 
+            destination=(1, 1), 
+            label_data=label_data, 
+            threshold=0.1,  # Much bigger than the 500m threshold, so we will get multiple bins
```

Comment:
@shankari 

this mock trip generator will randomly sample origins near the params for origin (0,0) and destination (1,1). it will _guarantee_ that some number of them are _within_threshold_, which by default is `None`. the remaining are guaranteed to be outside of that threshold. 

_BUT_, for those outside of the threshold, there is nothing preventing them from still being near each other. looking at the feature bins, i found this one:

```json
"30": {
        "features": [
            [
                0.007234893836876131,
                0.0016395550304311184,
                0.007234893836876131,
                0.0016395550304311184
            ],
            [
                0.003869586654049201,
                0.0017138992558370781,
                0.003869586654049201,
                0.0017138992558370781
            ]
        ],
        "labels": [ ... ],
        "predictions": [ ... ]
    },
```

i hopped over to jupyter and converted these linestrings to web mercator which is in a near-metric system projection, and these rows indeed appear to be within 500 meters of each other:


```
LINESTRING (805.385 182.514, 805.385 182.514)
LINESTRING (430.760 190.790, 430.760 190.790)
```

so now i'm not sure why the original algorithm didn't get this result.

Reply:
we could revise this mocked data in a similar way to your first backwards compat test building a matrix grid that we know has grid distance > similarity threshold

Reply:
> BUT, for those outside of the threshold, there is nothing preventing them from still being near each other. looking at the feature bins, i found this one:

True. As you can see from the comment, I expect to get multiple bins with multiple trips in each bin.
Not sure what the linestrings represent, but at least some of them are > 500m and some are on the cusp (430m) so maybe the near-metric is the problem?

After you re-run with the fix, how many bins do you get using the new method?

Reply:
ok so I re-ran with the fix and I get the same result

```
2022-08-09 16:37:01,436:DEBUG:old bins = 50 but new_bins = 35
```

which seems to indicate that maybe the fix didn't work...

Reply:
Re-tested the any versus all test and it does work

```
./e-mission-py.bash emission/tests/modellingTests/TestBackwardsCompat.py TestBackwardsCompat.testAnyVsAllWhilePredicting
----------------------------------------------------------------------
Ran 1 test in 0.021s

OK
```

Reply:
The monkeytest should still work if the implementation is the same since we are theoretically running the same algorithm on the same dataset. Let's add additional logging to see the differences...

Reply:
@robfitzgerald I added logs to the old similarity detection code

```
diff --git a/emission/analysis/modelling/tour_model/similarity.py b/emission/analysis/modelling/tour_model/similarity.py
index 58e1fecb..0ac2ddb0 100644
--- a/emission/analysis/modelling/tour_model/similarity.py
+++ b/emission/analysis/modelling/tour_model/similarity.py
@@ -39,6 +39,7 @@ This is called by cluster_pipeline.py.
 #Determine whether two points are within a cutoff distance threshold
 def within_radius(lat1, lon1, lat2, lon2, radius):
     dist = ecc.calDistance([lon1, lat1], [lon2, lat2])
+    print(f"Comparing distance {dist} to radius {radius}, returning {dist <= radius}")
     return dist <= radius

 def filter_too_short(all_trips, radius):
@@ -124,8 +125,8 @@ class similarity(object):
                         bin.append(a)
                         added = True
                         break
-                except:
-                    print(f"Got exception while matching trip {a}, creating new bin")
+                except Exception as e:
+                    print(f"Got exception {e} while matching trip {a}, creating new bin")
                     added = False
             if not added:
                 self.bins.append([a])
@@ -320,6 +321,7 @@ class similarity(object):
         enda = tripa.end_loc["coordinates"]
         endb = tripb.end_loc["coordinates"]

+        print(f"Comparing trip at index {a} and {b}")
         # Flip indices because points are in geojson (i.e. lon, lat)
         start = within_radius(starta[1], starta[0], startb[1], startb[0], self.radius)
         end = within_radius(enda[1], enda[0], endb[1], endb[0], self.radius)
```

And then ran the test (result in monkeytest_aug_9.log.gz).
And it turns out that there are some trips for which the start **or** the end are within the 500m radius, but in every case, the other side is outside the radius (start_end_true_aug_9.log.gz). Note that in the original implementation, we only return true from `withinDistance` if both start *and* end are within the radius.

[start_end_true_aug_9.log.gz](https://github.com/e-mission/e-mission-server/files/9295500/start_end_true_aug_9.log.gz)
[monkeytest_aug_9.log.gz](https://github.com/e-mission/e-mission-server/files/9295501/monkeytest_aug_9.log.gz)

Reply:
The regression in the new code is because the feature extraction is broken.

I added some additional logging to the new similarity metric

```
$ git diff emission/analysis/modelling/similarity/
diff --git a/emission/analysis/modelling/similarity/similarity_metric.py b/emission/analysis/modelling/similarity/similarity_metric.py
index 6be00216..59b8d70d 100644
--- a/emission/analysis/modelling/similarity/similarity_metric.py
+++ b/emission/analysis/modelling/similarity/similarity_metric.py
@@ -37,5 +37,6 @@ class SimilarityMetric(metaclass=ABCMeta):
         :return: true if the feature similarity is within some threshold
         """
         similarity_values = self.similarity(a, b)
+        print(f"while comparing features {a} and {b}, values = {similarity_values}")
         is_similar = all(map(lambda sim: sim <= thresh, similarity_values))
         return is_similar
```

And the start and end features are always identical, leading to identical start and end distances.

```
while comparing features [-0.0026629996435594883, 0.03876364351091465, -0.0026629996435594883, 0.0387636435109
1465] and [-0.02794919620160515, 0.016667362364481252, -0.02794919620160515, 0.016667362364481252], values = [3733.9601065577103, 3733.9601065577103]
while comparing features [0.008169998441481105, 0.007399881502856938, 0.008169998441481105, 0.0073998815028569
38] and [-0.02794919620160515, 0.016667362364481252, -0.02794919620160515, 0.016667362364481252], values = [4146.366766570032, 4146.366766570032]
while comparing features [0.008169998441481105, 0.007399881502856938, 0.008169998441481105, 0.0073998815028569
38] and [-0.0026629996435594883, 0.03876364351091465, -0.0026629996435594883, 0.03876364351091465], values = [3689.6604531600005, 3689.6604531600005]
...
```

Reply:
indeed you are right

```python
def destination_features(trip: ecwc.Confirmedtrip) -> List[float]:
    """extract the trip destination coordinates. 

    :param trip: trip to extract features from
    :return: destination coordinates
    """
    try:
        destination = trip['data']['start_loc']["coordinates"]
        return destination
    except KeyError as e:
        msg = 'Confirmedtrip expected to have path data.start_loc.coordinates'
        raise KeyError(msg) from e
```

fixing it now